### PR TITLE
Feature/restricted swap

### PR DESCRIPTION
**Pull request by voyager-gold**
https://github.com/CoMakery/comakery-security-token/issues/32

### CODE REVIEW

File: contracts/RestrictedSwap.sol [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    require(token2 != address(0), "Invalid token2 address");
+    
+    uint8 code = IERC1404(_erc1404).detectTransferRestriction(
+      restrictedTokenSender,
+      token2Sender,
+      restrictedTokenAmount);
+    string memory message = IERC1404(_erc1404).messageForTransferRestriction(code);
+    require(code == 0, message);
+
+    bytes memory data = abi.encodeWithSelector(
```
Comment (aquabu) [inline contracts/RestrictedSwap.sol:120]: why is `abi.encodeWithSelector` needed here?

File: contracts/RestrictedSwap.sol [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+  function fundRestrictedTokenSwap(uint256 swapNumber) external override {
+    Swap storage swap = _swap[swapNumber];
+    uint256 allowance = IERC20(_erc1404).allowance(msg.sender, address(this));
+
+    require(!swap.fundRestrictedToken, "This swap has already been funded");
+    require(!swap.canceled, "This swap has been canceled");
+    require(swap.restrictedTokenSender == msg.sender, "You are not appropriate token sender for this swap");
+    require(allowance >= swap.restrictedTokenAmount, "Insufficient allownace to transfer token");
+
+    uint256 balanceBefore = IERC20(_erc1404).balanceOf(address(this));
```
Comment (aquabu) [inline contracts/RestrictedSwap.sol:165]: Can lines `165 - 183` and `200 - 218` be factored into a single private method?

File: contracts/RestrictedSwap.sol [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+contract RestrictedSwap is IRestrictedSwap, AccessControl {
+  
+  struct Swap {
+    address restrictedTokenSender;
+    address token2Sender;
+    address token2;
+    uint256 restrictedTokenAmount;
+    uint256 token2Amount;
+    bool fundRestrictedToken;
```
Comment (aquabu) [inline contracts/RestrictedSwap.sol:19]: is this set when the the swap has been completed? Maybe could be named more clearly.

File: contracts/RestrictedSwap.sol (hunk 1 above)
Comment (voyager-gold) [reply contracts/RestrictedSwap.sol:120]: @aquabu This is for checking if token2 is erc20 or erc1404

File: contracts/RestrictedSwap.sol (hunk 3 above)
Comment (voyager-gold) [reply contracts/RestrictedSwap.sol:19]: Swap will be completed when both flags (fundRestrictedToken and fundToken2) are set true @aquabu

File: contracts/RestrictedSwap.sol (hunk 2 above)
Comment (voyager-gold) [reply contracts/RestrictedSwap.sol:165]: They look very similar but variables used are some what different.
How about leave them as now? :D @aquabu

**Pull request merged**